docs/modules/ROOT/pages/adr/0043-appcat-framework-production-architecture.adoc

```diff
@@ -0,0 +1,1750 @@
+kind: Composition
+metadata:
+  name: vshnredis.vshn.appcat.vshn.io
+spec:
+  mode: Pipeline
+  pipeline:
+  # Normal rendering
+  - step: redis-func
+    functionRef:
+      name: function-appcat
+    input:
+      kind: ConfigMap
+      name: appcat-redis-func-config
+
+  # Maintenance operation (runs periodically)
+  - step: maintenance-check
+    functionRef:
+      name: function-appcat-maintenance
+    operationRef:
+      name: maintenance-cron
+----
+
+**Advantages:**
+
+* *No Additional Dependencies*: Built into Crossplane
+* *Tightly Integrated*: Direct access to composed resources and claim state
+* *Declarative*: Same function model as composition
+
+**Disadvantages:**
+
+* *Experimental*: CronOperation/WatchOperation still alpha in Crossplane
+* *Limited Observability*: Logs only in function pod, no workflow UI
+* *Complex Logic in Go*: Maintenance logic embedded in composition function code
+* *No Multi-Step Workflows*: Can't express backup -> upgrade -> verify easily
+
+==== Decision
+
+Use CronJob pattern (Option A).
+Phase 2: Reevaluate Crossplane Operations (Option C) vs Argo Workflows (Option B) for complex orchestration such as release and rollback.
+
+==== Rationale
+
+1. *Proven Pattern*: CronJob-based maintenance works in component-appcat production.
+
+2. *Simplicity*: No additional dependencies, standard Kubernetes pattern.
+
+3. *Composition Revisions*: Leverage ADR0030 pattern (revision selector + automatic policy) for upgrades during maintenance windows.
+
+4. *Good Enough*: Version checks and revision updates don't require complex multi-step workflows yet.
+
+**Phase 2 Deferred Decision:**
+
+When CronJob maintenance becomes insufficient for complex multi-step workflows (backup → upgrade → verify → rollback):
+
+* *Option B (Argo Workflows)*: Production-ready workflow engine with DAG orchestration and dedicated UI
+* *Option C (Crossplane Operations)*: Native Crossplane integration but currently alpha status
+
+Reevaluate based on Crossplane Operations maturity, actual workflow complexity needs, and operational burden.
+
+=== 1.6: Testing Strategy
+
+==== Problem
+
+Need comprehensive testing strategy covering:
+
+* Composition function logic (Go code)
+* Service configuration validation (KCL)
+* Full service lifecycle (create, update, delete, backup, restore)
+* Regression testing for configuration changes
+
+==== Solutions
+
+===== Testing Layers
+
+====== Layer 1: Go Unit Tests (Composition Functions)
+
+**Tool:** Go testing framework (`testing` package)
+
+**Coverage:**
+
+* Composition function logic
+* Helm value rendering
+* ConfigMap loading and parsing
+* Connection detail handling
+* All other managed resources during composition function life-cycle
+
+**Advantages:**
+
+* *Fast*: No cluster needed, runs in milliseconds
+* *Focused*: Tests specific logic units
+* *Easy to Debug*: Direct Go debugging with breakpoints
+* *CI-Friendly*: Quick feedback in pull requests
```

Comment:
Disadvantages:
- Need cumbersome testing fixtures to emulate Crossplane grpc requests

Of course we can put in work to abstract that away. But that's then still additional effort.